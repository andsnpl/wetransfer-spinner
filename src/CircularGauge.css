.gauge {
  display: block;
  position: relative;
  border-radius: 50%;
  overflow: hidden;
  width: 232px;
  height: 232px;
}

.gauge-circle {
  width: 100%;
  height: 100%;
}

.gauge-background-panel {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 50%;
  left: 50%;
  /* transform: property to be applied as an inline style */
  transform-origin: bottom left;
  transition: transform 0.1s linear;
}

.gauge-endcap-container {
  position: absolute;
  top: 0;
  right: 50%;
  left: 50%;
  bottom: 50%;
  /* transform: property to be applied as an inline style */
  transform-origin: bottom center;
  transition: transform 0.1s linear;
}

.gauge-endcap {
  display: block;
  position: absolute;
  top: 0;
  left: 50%;
  width: 12px;
  height: 12px;
  margin-left: -6px;
  border-radius: 50%;
  /* transform: property to be applied as an inline style */
  transition: transform 0.1s linear;
}

.gauge-center {
  position: absolute;
  inset: 12px;
  border-radius: 50%;
  display: flex;
  align-items: center;
  justify-content: center;
}

.gauge-value {
  font-size: 88px;
  line-height: 60px;
}

.gauge-percent-sign {
  font-size: 24px;
  line-height: 24px;
  color: #babcbf;
  font-weight: bold;
}

.gauge-value,
.gauge-percent-sign {
  vertical-align: top;
}
