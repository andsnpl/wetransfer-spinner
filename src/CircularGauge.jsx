import "./CircularGauge.css";

function GaugeCircle({ color, origin, children }) {
  const rotationCss = origin && `rotate(${origin})`;
  return (
    <div
      className="gauge-circle"
      style={{
        backgroundColor: color,
        transform: rotationCss
      }}
      aria-hidden
    >
      {children}
    </div>
  );
}

function GaugeBackgroundPanel({ color, rotation }) {
  const rotationCss = `rotate(${rotation}deg)`;
  return (
    <div
      className="gauge-background-panel"
      style={{
        backgroundColor: color,
        transform: rotationCss
      }}
    />
  );
}

function GaugeEndcap({ color, rotation, isVisible }) {
  const rotationCss = `rotate(${rotation}deg)`;
  const scaleCss = isVisible ? "scaleX(1)" : "scaleX(0)";
  return (
    <div className="gauge-endcap-container" style={{ transform: rotationCss }}>
      <div
        className="gauge-endcap"
        style={{ transform: scaleCss, backgroundColor: color }}
      />
    </div>
  );
}

function GaugeReadout({ value }) {
  if (typeof value === "number") {
    value = value.toFixed(0);
  }
  return (
    <div className="gauge-readout">
      <span className="gauge-value">{value}</span>
      {value && <span className="gauge-percent-sign">%</span>}
    </div>
  );
}

function CircularGauge({
  value,
  origin = null,
  backgroundColor: bg = "#e2e4e6",
  fillColor: fg = "#409fff",
  pageColor = "#ffffff",
  children = null
}) {
  value = Math.max(Math.min(value, 100), 0);
  const arc = (value * 360) / 100;
  return (
    <div className="gauge">
      <GaugeCircle color={fg} origin={origin}>
        {/**
         * Four background panels covering 90deg each that rotate out of the
         * way to reveal the circle color + one stationary panel filled with
         * the fill color, to allow the background panels to be completely
         * hidden at 100%.
         */}
        <GaugeBackgroundPanel color={bg} rotation={Math.max(arc, 0)} />
        <GaugeBackgroundPanel color={bg} rotation={Math.max(arc, 90)} />
        <GaugeBackgroundPanel color={bg} rotation={Math.max(arc, 180)} />
        <GaugeBackgroundPanel color={bg} rotation={Math.max(arc, 270)} />
        {arc > 90 && (
          <GaugeBackgroundPanel color={fg} rotation={0} minRotation={0} />
        )}
        <GaugeEndcap color={fg} rotation={0} isVisible={arc > 0} />
        <GaugeEndcap color={fg} rotation={arc} isVisible={arc > 0} />
      </GaugeCircle>
      <div className="gauge-center" style={{ backgroundColor: pageColor }}>
        {children ? children : <GaugeReadout value={value} />}
      </div>
    </div>
  );
}

// Export the readout component as a convenience, to help callers override
// the display.
CircularGauge.Readout = GaugeReadout;

export default CircularGauge;
