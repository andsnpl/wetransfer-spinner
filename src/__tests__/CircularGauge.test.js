import { render } from "@testing-library/react";
import CircularGauge from "../CircularGauge";

describe("CircularGauge", () => {
  it("should render with 0%", () => {
    const { container } = render(<CircularGauge value={0} />);
    expect(container).toMatchSnapshot();
  });

  it("should render with 1%", () => {
    const { container } = render(<CircularGauge value={1} />);
    expect(container).toMatchSnapshot();
  });

  it("should render with 1% and an offset origin", () => {
    const { container } = render(<CircularGauge value={1} origin={30} />);
    expect(container).toMatchSnapshot();
  });

  it("should render with 50%", () => {
    const { container } = render(<CircularGauge value={1} />);
    expect(container).toMatchSnapshot();
  });

  it("should render with 100%", () => {
    const { container } = render(<CircularGauge value={100} />);
    expect(container).toMatchSnapshot();
  });

  it("should render with custom colors", () => {
    const { container } = render(
      <CircularGauge
        value={10}
        fillColor="#bada55"
        backgroundColor="#00b1ec"
        pageColor="#1e1e1e"
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("should render with custom children", () => {
    const { container } = render(
      <CircularGauge value={77}>
        Custom <strong>children</strong>
      </CircularGauge>
    );
    expect(container).toMatchSnapshot();
  });
});
