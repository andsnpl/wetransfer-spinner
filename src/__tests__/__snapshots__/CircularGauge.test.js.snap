// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CircularGauge should render with 0% 1`] = `
<div>
  <div
    class="gauge"
  >
    <div
      aria-hidden="true"
      class="gauge-circle"
      style="background-color: rgb(64, 159, 255);"
    >
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(0deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(90deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(180deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(270deg);"
      />
      <div
        class="gauge-endcap-container"
        style="transform: rotate(0deg);"
      >
        <div
          class="gauge-endcap"
          style="transform: scaleX(0); background-color: rgb(64, 159, 255);"
        />
      </div>
      <div
        class="gauge-endcap-container"
        style="transform: rotate(0deg);"
      >
        <div
          class="gauge-endcap"
          style="transform: scaleX(0); background-color: rgb(64, 159, 255);"
        />
      </div>
    </div>
    <div
      class="gauge-center"
      style="background-color: rgb(255, 255, 255);"
    >
      <div
        class="gauge-readout"
      >
        <span
          class="gauge-value"
        >
          0
        </span>
        <span
          class="gauge-percent-sign"
        >
          %
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`CircularGauge should render with 1% 1`] = `
<div>
  <div
    class="gauge"
  >
    <div
      aria-hidden="true"
      class="gauge-circle"
      style="background-color: rgb(64, 159, 255);"
    >
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(3.6deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(90deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(180deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(270deg);"
      />
      <div
        class="gauge-endcap-container"
        style="transform: rotate(0deg);"
      >
        <div
          class="gauge-endcap"
          style="transform: scaleX(1); background-color: rgb(64, 159, 255);"
        />
      </div>
      <div
        class="gauge-endcap-container"
        style="transform: rotate(3.6deg);"
      >
        <div
          class="gauge-endcap"
          style="transform: scaleX(1); background-color: rgb(64, 159, 255);"
        />
      </div>
    </div>
    <div
      class="gauge-center"
      style="background-color: rgb(255, 255, 255);"
    >
      <div
        class="gauge-readout"
      >
        <span
          class="gauge-value"
        >
          1
        </span>
        <span
          class="gauge-percent-sign"
        >
          %
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`CircularGauge should render with 1% and an offset origin 1`] = `
<div>
  <div
    class="gauge"
  >
    <div
      aria-hidden="true"
      class="gauge-circle"
      style="background-color: rgb(64, 159, 255); transform: rotate(30);"
    >
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(3.6deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(90deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(180deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(270deg);"
      />
      <div
        class="gauge-endcap-container"
        style="transform: rotate(0deg);"
      >
        <div
          class="gauge-endcap"
          style="transform: scaleX(1); background-color: rgb(64, 159, 255);"
        />
      </div>
      <div
        class="gauge-endcap-container"
        style="transform: rotate(3.6deg);"
      >
        <div
          class="gauge-endcap"
          style="transform: scaleX(1); background-color: rgb(64, 159, 255);"
        />
      </div>
    </div>
    <div
      class="gauge-center"
      style="background-color: rgb(255, 255, 255);"
    >
      <div
        class="gauge-readout"
      >
        <span
          class="gauge-value"
        >
          1
        </span>
        <span
          class="gauge-percent-sign"
        >
          %
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`CircularGauge should render with 50% 1`] = `
<div>
  <div
    class="gauge"
  >
    <div
      aria-hidden="true"
      class="gauge-circle"
      style="background-color: rgb(64, 159, 255);"
    >
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(3.6deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(90deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(180deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(270deg);"
      />
      <div
        class="gauge-endcap-container"
        style="transform: rotate(0deg);"
      >
        <div
          class="gauge-endcap"
          style="transform: scaleX(1); background-color: rgb(64, 159, 255);"
        />
      </div>
      <div
        class="gauge-endcap-container"
        style="transform: rotate(3.6deg);"
      >
        <div
          class="gauge-endcap"
          style="transform: scaleX(1); background-color: rgb(64, 159, 255);"
        />
      </div>
    </div>
    <div
      class="gauge-center"
      style="background-color: rgb(255, 255, 255);"
    >
      <div
        class="gauge-readout"
      >
        <span
          class="gauge-value"
        >
          1
        </span>
        <span
          class="gauge-percent-sign"
        >
          %
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`CircularGauge should render with 100% 1`] = `
<div>
  <div
    class="gauge"
  >
    <div
      aria-hidden="true"
      class="gauge-circle"
      style="background-color: rgb(64, 159, 255);"
    >
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(360deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(360deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(360deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(360deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(64, 159, 255); transform: rotate(0deg);"
      />
      <div
        class="gauge-endcap-container"
        style="transform: rotate(0deg);"
      >
        <div
          class="gauge-endcap"
          style="transform: scaleX(1); background-color: rgb(64, 159, 255);"
        />
      </div>
      <div
        class="gauge-endcap-container"
        style="transform: rotate(360deg);"
      >
        <div
          class="gauge-endcap"
          style="transform: scaleX(1); background-color: rgb(64, 159, 255);"
        />
      </div>
    </div>
    <div
      class="gauge-center"
      style="background-color: rgb(255, 255, 255);"
    >
      <div
        class="gauge-readout"
      >
        <span
          class="gauge-value"
        >
          100
        </span>
        <span
          class="gauge-percent-sign"
        >
          %
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`CircularGauge should render with custom children 1`] = `
<div>
  <div
    class="gauge"
  >
    <div
      aria-hidden="true"
      class="gauge-circle"
      style="background-color: rgb(64, 159, 255);"
    >
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(277.2deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(277.2deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(277.2deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(226, 228, 230); transform: rotate(277.2deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(64, 159, 255); transform: rotate(0deg);"
      />
      <div
        class="gauge-endcap-container"
        style="transform: rotate(0deg);"
      >
        <div
          class="gauge-endcap"
          style="transform: scaleX(1); background-color: rgb(64, 159, 255);"
        />
      </div>
      <div
        class="gauge-endcap-container"
        style="transform: rotate(277.2deg);"
      >
        <div
          class="gauge-endcap"
          style="transform: scaleX(1); background-color: rgb(64, 159, 255);"
        />
      </div>
    </div>
    <div
      class="gauge-center"
      style="background-color: rgb(255, 255, 255);"
    >
      Custom 
      <strong>
        children
      </strong>
    </div>
  </div>
</div>
`;

exports[`CircularGauge should render with custom colors 1`] = `
<div>
  <div
    class="gauge"
  >
    <div
      aria-hidden="true"
      class="gauge-circle"
      style="background-color: rgb(186, 218, 85);"
    >
      <div
        class="gauge-background-panel"
        style="background-color: rgb(0, 177, 236); transform: rotate(36deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(0, 177, 236); transform: rotate(90deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(0, 177, 236); transform: rotate(180deg);"
      />
      <div
        class="gauge-background-panel"
        style="background-color: rgb(0, 177, 236); transform: rotate(270deg);"
      />
      <div
        class="gauge-endcap-container"
        style="transform: rotate(0deg);"
      >
        <div
          class="gauge-endcap"
          style="transform: scaleX(1); background-color: rgb(186, 218, 85);"
        />
      </div>
      <div
        class="gauge-endcap-container"
        style="transform: rotate(36deg);"
      >
        <div
          class="gauge-endcap"
          style="transform: scaleX(1); background-color: rgb(186, 218, 85);"
        />
      </div>
    </div>
    <div
      class="gauge-center"
      style="background-color: rgb(30, 30, 30);"
    >
      <div
        class="gauge-readout"
      >
        <span
          class="gauge-value"
        >
          10
        </span>
        <span
          class="gauge-percent-sign"
        >
          %
        </span>
      </div>
    </div>
  </div>
</div>
`;
